src/v/iceberg/datatypes.cc

```diff
@@ -0,0 +1,279 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/datatypes.h"
+
+#include <variant>
+
+namespace iceberg {
+
+struct primitive_type_comparison_visitor {
+    template<typename T, typename U>
+    bool operator()(const T&, const U&) const {
+        return false;
+    }
+    bool operator()(const boolean_type&, const boolean_type&) const {
```

Comment:
Ah true, done. Was hesitant because of decimals and fixed, but the compiler is smart enough to use the right one when available.

If other types are introduced though, having a default may be error prone. But for now it is a nice cleanup